File: app/api/quiz.py
import json

from fastapi import APIRouter, Depends, HTTPException
from sqlalchemy.orm import Session

from app.db.sqlite import get_db
from app.db import crud
from app.services import groq_client
from app.schemas.quiz import (
    QuizGenerateRequest,
    QuizGenerateResponse,
    QuizQuestion,
    QuizOption,
    QuizSubmitRequest,
    QuizSubmitResponse,
    QuizAnswerBreakdown,
)

router = APIRouter(prefix="/api/quiz", tags=["quiz"])


@router.post("/generate", response_model=QuizGenerateResponse)
def generate_quiz(request: QuizGenerateRequest, db: Session = Depends(get_db)):
    article = crud.get_article(db, request.article_id)
    if not article:
        raise HTTPException(status_code=404, detail="Article not found")

    raw_questions = groq_client.generate_quiz(article.summary)
    if not raw_questions:
        raise HTTPException(status_code=422, detail="LLM failed to generate quiz questions")

    session = crud.create_quiz_session(
        db,
        user_id=request.user_id,
        article_id=request.article_id,
        domain=article.domain,
        questions=raw_questions,
    )

    questions_out = []
    for q in raw_questions:
        opts = q.get("options", {})
        questions_out.append(QuizQuestion(
            question=q["question"],
            options=QuizOption(
                A=opts.get("A", ""),
                B=opts.get("B", ""),
                C=opts.get("C", ""),
                D=opts.get("D", ""),
            ),
            explanation=q.get("explanation"),
        ))

    return QuizGenerateResponse(
        session_id=session.id,
        article_id=request.article_id,
        domain=article.domain,
        questions=questions_out,
    )


@router.post("/submit", response_model=QuizSubmitResponse)
def submit_quiz(request: QuizSubmitRequest, db: Session = Depends(get_db)):
    session = crud.get_quiz_session(db, request.session_id)
    if not session:
        raise HTTPException(status_code=404, detail="Quiz session not found")
    if session.submitted_at is not None:
        raise HTTPException(status_code=409, detail="Quiz already submitted")

    updated = crud.submit_quiz_session(db, request.session_id, request.answers)
    questions = json.loads(updated.questions_json)

    breakdown = []
    for i, q in enumerate(questions):
        user_ans = request.answers.get(str(i), "").upper()
        correct_ans = q["answer"].upper()
        breakdown.append(QuizAnswerBreakdown(
            question=q["question"],
            your_answer=user_ans,
            correct_answer=correct_ans,
            is_correct=user_ans == correct_ans,
            explanation=q.get("explanation", ""),
        ))

    return QuizSubmitResponse(
        session_id=updated.id,
        score=round(updated.score, 3),
        correct_count=updated.correct_count,
        total_questions=updated.total_questions,
        breakdown=breakdown,
    )

File: frontend/lib/core/models/quiz.dart
class QuizQuestion {
  final int index;
  final String question;
  final Map<String, String> options; // {A: "...", B: "...", C: "...", D: "..."}
  final String? explanation;

  const QuizQuestion({
    required this.index,
    required this.question,
    required this.options,
    this.explanation,
  });

  factory QuizQuestion.fromJson(int index, Map<String, dynamic> json) {
    final rawOpts = json['options'] as Map<String, dynamic>;
    return QuizQuestion(
      index: index,
      question: json['question'] as String,
      options: rawOpts.map((k, v) => MapEntry(k, v as String)),
      explanation: json['explanation'] as String?,
    );
  }

  List<String> get optionLabels => options.keys.toList()..sort();
}

class QuizSession {
  final int sessionId;
  final int articleId;
  final String domain;
  final List<QuizQuestion> questions;

  const QuizSession({
    required this.sessionId,
    required this.articleId,
    required this.domain,
    required this.questions,
  });

  factory QuizSession.fromJson(Map<String, dynamic> json) {
    final rawQuestions = json['questions'] as List;
    return QuizSession(
      sessionId: json['session_id'] as int,
      articleId: json['article_id'] as int,
      domain: json['domain'] as String? ?? '',
      questions: rawQuestions
          .asMap()
          .entries
          .map((e) => QuizQuestion.fromJson(
              e.key, e.value as Map<String, dynamic>))
          .toList(),
    );
  }
}

class QuizAnswerBreakdown {
  final String question;
  final String yourAnswer;
  final String correctAnswer;
  final bool isCorrect;
  final String explanation;

  const QuizAnswerBreakdown({
    required this.question,
    required this.yourAnswer,
    required this.correctAnswer,
    required this.isCorrect,
    required this.explanation,
  });

  factory QuizAnswerBreakdown.fromJson(Map<String, dynamic> json) =>
      QuizAnswerBreakdown(
        question: json['question'] as String,
        yourAnswer: json['your_answer'] as String,
        correctAnswer: json['correct_answer'] as String,
        isCorrect: json['is_correct'] as bool,
        explanation: json['explanation'] as String,
      );
}

class QuizResult {
  final int sessionId;
  final double score;
  final int correctCount;
  final int totalQuestions;
  final List<QuizAnswerBreakdown> breakdown;

  const QuizResult({
    required this.sessionId,
    required this.score,
    required this.correctCount,
    required this.totalQuestions,
    required this.breakdown,
  });

  factory QuizResult.fromJson(Map<String, dynamic> json) => QuizResult(
        sessionId: json['session_id'] as int,
        score: (json['score'] as num).toDouble(),
        correctCount: json['correct_count'] as int,
        totalQuestions: json['total_questions'] as int,
        breakdown: (json['breakdown'] as List)
            .map((e) =>
                QuizAnswerBreakdown.fromJson(e as Map<String, dynamic>))
            .toList(),
      );

  double get percentage =>
      totalQuestions == 0 ? 0 : correctCount / totalQuestions;
}
